infrastructure/src/main/java/org/corfudb/infrastructure/log/Segment.java

```diff
@@ -0,0 +1,557 @@
+            channel.position(0);
+            return null;
+        }
+
+        ByteBuffer buffer = getPayloadForMetadata(channel, metadata);
+        if (buffer == null) {
+            // partial write on the header payload
+            // Rewind the channel position to the beginning of the file
+            channel.position(0);
+            return null;
+        }
+
+        if (Checksum.getChecksum(buffer.array()) != metadata.getPayloadChecksum()) {
+            String errorMessage = getDataCorruptionErrorMessage("Invalid metadata checksum",
+                    channel, segmentFilePath
+            );
+            throw new DataCorruptionException(errorMessage);
+        }
+
+        LogFormat.LogHeader header;
+
+        try {
+            header = LogFormat.LogHeader.parseFrom(buffer.array());
+        } catch (InvalidProtocolBufferException e) {
+            String errorMessage = getDataCorruptionErrorMessage("Invalid header",
+                    channel, segmentFilePath
+            );
+            throw new DataCorruptionException(errorMessage, e);
+        }
+
+        return header;
+    }
+
+    public synchronized void retain() {
+        refCount++;
+    }
+
+    public synchronized void release() {
+        if (refCount == 0) {
+            throw new IllegalStateException("refCount cannot be less than 0, segment " + id);
+        }
+        refCount--;
+    }
+
+    public void close() {
+
+        Set<FileChannel> channels = new HashSet<>(
+                Arrays.asList(writeChannel, readChannel)
+        );
+
+        for (FileChannel channel : channels) {
+            try {
+                channel.force(true);
+            } catch (IOException e) {
+                log.error("Can't force updates for {}", segmentFilePath, e);
+            } finally {
+                IOUtils.closeQuietly(channel);
+            }
+        }
+
+        if (refCount != 0) {
+            log.warn("closeSegmentHandlers: Segment {} is trimmed, but refCount is {}, attempting to trim anyways",
+                    segmentFilePath, refCount);
+        }
+    }
+
+    /**
+     * This index maps a sequence number to a file offset + payload length.
+     *
+     * The file offset and payload length are packed in a long (treated as unsigned) as follows:
+     *
+     *  |--------64-bits--------|
+     *  |fileOffset|payloadSize|
+     *
+     */
+    private static final class Index {
+
+        private final BoundedMap map;
+        private final int highBitsNum;
+
+        private Index(long offset, int size) {
+            this.map = new BoundedMap(offset, size);
+            this.highBitsNum = Long.bitCount(MAX_SEGMENT_SIZE);
+            int totalEncodedBytes = this.highBitsNum + Integer.bitCount(MAX_WRITE_SIZE) >> 3;
+            Preconditions.checkArgument(totalEncodedBytes == Long.BYTES);
+        }
+
+        private long pack(long high, int low) {
+            return (((long) high) << this.highBitsNum) | (low & MAX_WRITE_SIZE);
```

Comment:
Do we need Precondition check on the given high and low? although it is extremely unlikely.

Reply:
Yes, but not here. 

We should add the precondition to `Segment::write` that way the write is failed before it reaches the filesystem. The indexes are only updated after the write is sent to the file system.